File: advent_of_code/day6/part2/main.py
from collections import defaultdict
from typing import DefaultDict, Dict, List

from advent_of_code.helpers.helpers import read_file


def compile_groups(line: str) -> List[List[str]]:
    return [i.split("\n") for i in line.split("\n\n")] 


def group_data(group: List[str]) -> Dict[str, int]:
    map: DefaultDict[str, int] = defaultdict(int)
    map["count"] = len(group)
    for g in group:
        for letter in list(g):
            map[letter] += 1
    return dict(map)


def get_count_from_group(group: Dict[str, int]) -> int:
    return list(group.values()).count(group.get("count")) - 1


def main(filepath: str) -> int:
    lines: str = read_file(filepath=filepath)
    groups: List[str] = compile_groups(lines)
    grouped_data: List[Dict[str, int]] = [group_data(g) for g in groups]
    return sum([get_count_from_group(gg) for gg in grouped_data])

File: advent_of_code/day6/part2/main.go
package part2

import (
	"advent-of-code/advent_of_code/helpers"
	"strings"
)

func compileGroupsV2(data string) [][]string {
	groups := make([][]string, 0)
	newData := strings.Split(data, "\n\n")
	for _, i := range newData {
		group := make([]string, 0)
		temp := strings.Split(i, "\n")
		for _, j := range temp {
			group = append(group, j)
		}
		groups = append(groups, group)
	}
	return groups
}

func groupData(group []string) map[string]int {
	set := make(map[string]int)
	set["count"] = len(group)
	for _, p := range group {
		letters := strings.Split(p, "")
		for _, c := range letters {
			set[c] += 1
		}
	}
	return set
}

func getCountFromGroup(set map[string]int) int {
	count := 0
	for _, v := range set {
		if v == set["count"] {
			count++
		}
	}
	return count - 1
}

func Main(filepath string) int {

	file, err := helpers.ReadFile(filepath)
	if err != nil {
		panic(err)
	}
	groups := compileGroupsV2(file)
	count := 0

	for _, g := range groups {
		set := groupData(g)
		count += getCountFromGroup(set)
	}

	return count

}
